spicy/toolchain/include/compiler/detail/codegen/productions/while.h

```diff
@@ -8,33 +8,70 @@
  */
 class While : public ProductionBase, public spicy::trait::isNonTerminal {
 public:
+    /**
+     * Constructor for a while-loop using an expression as the condition for termination.
+     */
     While(const std::string& symbol, Expression e, Production body, const Location& l = location::None)
-        : ProductionBase(symbol, l), _expression(std::move(e)), _body(std::move(body)) {}
+        : ProductionBase(symbol, l), _body(std::move(body)), _expression(std::move(e)) {}
 
-    const Expression& expression() const { return _expression; }
-    const Production& body() const { return _body; }
+    /**
+     * Constructor for a while-loop using look-ahead as the condition for
+     * termination. When using this constructor, `preprocessLookAhead()` must
+     * later be called with the grammar that the production has been inserted
+     * into.
+     */
+    While(const std::string& symbol, Production body, const Location& l =
+            location::None);
+
+    /** Returns the loop expression if passed into the corresponding constructor. */
+    const auto& expression() const { return _expression; }
+
+    /** Returns the body production as passed into any of the constructors.  */
+    const auto& body() const { return _body; }
+
+    /**
+     * Prepares the internal grammar representation for a look-ahead based
+     * loop. Must be called (only) when the corresponding constructor was used.
+     *
+     * @param grammar grammar that while-production is being part of.
+     */
+    void preprocessLookAhead(Grammar* grammar);
+
+    /**
+     * For a look-ahead loop, returns the internally generated `LookAhead`
+     * production that's being used for generating the code to terminate the
+     * loop. The production's 1st alternative corresponds to the case of
+     * terminating the loop, the 2nd alternative corresponds to executing the
+     * loop body. This method must be called only after `preprocessLookAhead()`.
+     */
+    const production::LookAhead& lookAheadProduction() const {
+        assert(_body_for_grammar); // set by preprocessLookAhead() return
+        return _body_for_grammar->as<production::LookAhead>();
+    }
 
     // Production API
-    std::vector<std::vector<Production>> rhss() const { return {{_body}}; }
+    std::vector<std::vector<Production>> rhss() const { return {{(_body_for_grammar ? *_body_for_grammar : _body)}}; }
     std::optional<spicy::Type> type() const { return {}; }
     bool nullable() const { return production::nullable(rhss()); }
     bool eodOk() const { return nullable(); }
     bool atomic() const { return false; }
     bool supportsSynchronize() const { return hasSize() || _body.supportsSynchronize(); }
-    std::string render() const { return hilti::util::fmt("while(%s): %s", _expression, _body.symbol()); }
+    std::string render() const;
 
 private:
-    Expression _expression;
     Production _body;
+    std::optional<Expression> _expression;
```

Comment:
Let's add a comment here that only one of these should be set.

Reply:
Actually `_body` is always set.

Reply:
I meant that only one of `_expression` or `_body_for_grammar` should be set making this some sort of union type.